pkg/virt-api/webhooks/mutating-webhook/mutators/vm-mutator_test.go

```diff
@@ -350,6 +367,778 @@ var _ = Describe("VirtualMachine Mutator", func() {
+					Source: &v1beta1.DataVolumeSource{
+						PVC: &v1beta1.DataVolumeSourcePVC{
+							Name:      pvc.Name,
+							Namespace: pvc.Namespace,
+						},
+					},
+				},
+			}
+			dvWithSourcePVC, err = virtClient.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Create(context.Background(), dvWithSourcePVC, k8smetav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			dsWithSourcePVC = &v1beta1.DataSource{
+				ObjectMeta: k8smetav1.ObjectMeta{
+					Name:      dsWithSourcePVCName,
+					Namespace: vm.Namespace,
+				},
+				Spec: v1beta1.DataSourceSpec{
+					Source: v1beta1.DataSourceSource{
+						PVC: &v1beta1.DataVolumeSourcePVC{
+							Name:      pvc.Name,
+							Namespace: pvc.Namespace,
+						},
+					},
+				},
+			}
+			dsWithSourcePVC, err = virtClient.CdiClient().CdiV1beta1().DataSources(vm.Namespace).Create(context.Background(), dsWithSourcePVC, k8smetav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			dsWithAnnotations = &v1beta1.DataSource{
+				ObjectMeta: k8smetav1.ObjectMeta{
+					Name:      dsWithAnnotionsName,
+					Namespace: vm.Namespace,
+					Annotations: map[string]string{
+						apiinstancetype.DefaultInstancetypeAnnotation:     defaultInferedName,
+						apiinstancetype.DefaultInstancetypeKindAnnotation: defaultInferedKind,
+						apiinstancetype.DefaultPreferenceAnnotation:       defaultInferedName,
+						apiinstancetype.DefaultPreferenceKindAnnotation:   defaultInferedKind,
+					},
+				},
+				TypeMeta: k8smetav1.TypeMeta{
+					Kind: "DataSource",
+				},
+				Spec: v1beta1.DataSourceSpec{
+					Source: v1beta1.DataSourceSource{
+						PVC: &v1beta1.DataVolumeSourcePVC{
+							Name:      pvc.Name,
+							Namespace: pvc.Namespace,
+						},
+					},
+				},
+			}
+			_, err = virtClient.CdiClient().CdiV1beta1().DataSources(vm.Namespace).Create(context.Background(), dsWithAnnotations, k8smetav1.CreateOptions{})
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		DescribeTable("should infer defaults from VolumeSource and PersistentVolumeClaim", func(instancetypeMatcher, expectedInstancetypeMatcher *v1.InstancetypeMatcher, preferenceMatcher, expectedPreferenceMatcher *v1.PreferenceMatcher) {
+			vm.Spec.Instancetype = instancetypeMatcher
+			vm.Spec.Preference = preferenceMatcher
+			vm.Spec.Template.Spec.Volumes = []v1.Volume{{
+				Name: inferVolumeName,
+				VolumeSource: v1.VolumeSource{
+					PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+						PersistentVolumeClaimVolumeSource: k8sv1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.Name,
+						},
+					},
+				},
+			}}
+			vmSpec, _ := getVMSpecMetaFromResponse()
+			Expect(vmSpec.Instancetype).To(Equal(expectedInstancetypeMatcher))
+			Expect(vmSpec.Preference).To(Equal(expectedPreferenceMatcher))
+		},
+			Entry("for InstancetypeMatcher",
+				&v1.InstancetypeMatcher{
+					InferFromVolume: inferVolumeName,
+				},
+				&v1.InstancetypeMatcher{
+					Name: defaultInferedName,
```

Comment:
See the comment about annotation values.

Reply:
ACK done.